### DISCUSSION TITLE
How to access/compute second order derivates

### DISCUSSION THREAD

**Discussion by ABallisat** [idaholab/moose#17183] [Q&A Modules: General] [answered]
Hello all,

I am trying to implement a surface tension term within the level set module which requires computing the curvature of the interface. Explicitly, I need to calculate (ignoring constants):

F = (grad dot grad phi)/|grad phi| * grad phi

for the level set variable phi. I believe I cannot get around the grad dot grad phi term when putting it in the weak form, and I cannot work out how to access it or compute it in the kernel. Is there an easy way in MOOSE to access the second order derivates?

Thanks in advance for the help,

Alex

**Comment by hugary1995** [comment, marked as answer]
If I were you I would first try really hard to see if I can "shift" the derivative onto the test function in the weak form. But in case you really can't:

See [https://github.com/idaholab/moose/blob/next/modules/phase_field/include/kernels/CHInterfaceBase.h](https://github.com/idaholab/moose/blob/next/modules/phase_field/include/kernels/CHInterfaceBase.h) for example how to get second spatial derivatives of variable, test and trial functions.

But you need higher continuity, e.g. a third order hermite polynomial as your basis.

Alternatively, you can try splitting up your weak form by introducing a new variable (just like the split form cahn hilliard equation).

**Reply by ABallisat** [reply]
Thanks for the advice. I will focus for now on trying to shift the derivative onto the test function and if I can't get around it I will go into those suggestions.

**Reply by jiangwen84** [reply]
Take a look at Elin Olsson et.al, JCP 225 (2007) 785–807 paper where he has a formulation for regularized curvature. I suggest introduce a kernel to compute the curvature variable based on that formulation.

I have implemented it in our internal MOOSE-based codes, and it works well.

**Reply by ABallisat** [reply]
Thanks for that, I will take a good look and give that approach a go. Just out of interest is there any plan to incorporate that kernel into the general MOOSE distribution?

**Reply by jiangwen84** [reply]
Yes, those codes will be eventually merged into MOOSE, maybe as a new physical module. However, the timeline is not very clear at this moment.

**Reply by ABallisat** [reply]
I'm having trouble getting this to give sensible answers. For clarity, I am implementing equation 22 of that paper, using three kernels, one for each of the three integral terms and moving them all onto the same side so flipping some of the signs to get the correct residual calculations. Apply Neumann BCs to the curvature variable and level set variable (I am using the smeared heaviside function as per the first paper in that series). At the moment this is giving the following, where kappa is the curvature and k_Magnitude is the magnitude of the of gradient of the level set function (which is simply a stationary circle centred on 0.1 with radius 0.025). I know the gradient is correct but as the curvature should be the gradient of gradient, this does not seem to be correct. Am I missing something obvious or did you have to do anything in particular to get it working? I can't see what I am doing wrong...

Thanks for the help!

![image](https://user-images.githubusercontent.com/40231544/111353139-e6223400-867c-11eb-859d-ea53187e8449.png)

**Reply by jiangwen84** [reply]
Curvature = div (normal) = div (grad_phi / ||grad_phi||)

for signed distance function ||grad_phi||=1

so Curvature = div grad_phi = Lap phi.

**Reply by ABallisat** [reply]
Ah that solved it, I think my understanding of the reinitialisation and the conversion to the smeared heaviside function was incorrect. Thanks again for all the help, it is greatly appreciated!

**Reply by alimostafavi24** [reply]
> Curvature = div (normal) = div (grad_phi / ||grad_phi||)
> 
> for signed distance function ||grad_phi||=1
> 
> so Curvature = div grad_phi = Lap phi.

Hello,

Here as you suggested I used a custom kernel to compute the curvature. Bu this really messed up with the speed and convergence of my model. Is there any Auxkernel that can be used here and get the curvature as an Auxvariable? and curvature is simply the laplacian of phi?

Thanks,

**Reply by GiudGiud** [reply]
`curvature as an Auxvariable` 
you cannot use this (without extra work) for computing the residuals though. Auxvariables do NOT carry derivative information for the Jacobian

But if you really want to just look at it, in the output, then you ll have to make a new auxkernel, we dont have that one.
I can make it too if you have a week or two to wait

**Reply by lindsayad** [reply]
You may get curvatures from FE computation directly in libMesh: [get_curvatures](https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1FEAbstract.html#a7b50fa327c33945c6fa3bfd2da230cde). As alluded to by @jiangwen84 it does require pre-requesting [get_d2phi](https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1FEGenericBase.html#a2b469a4bf1136ba8fefb1e32c7fae991)

**Reply by lindsayad** [reply]
Note that we also have an `adCurvatures` method in `Assembly`, which I am fixing in https://github.com/idaholab/moose/pull/24204/commits/e29ac6b255 . We could create a non-AD curvatures method as well or you can just use existing second derivative machinery

**Reply by lindsayad** [reply]
Getting into the weeds here, but here is how we calculate curvatures in 2D in libMesh:
```c++
                // The curvature is computed via the familiar Frenet formula:
                // curvature = [d^2(x) / d (xi)^2] dot [normal]
                // For a reference, see:
                // F.S. Merritt, Mathematics Manual, 1962, McGraw-Hill, p. 310
                //
                // Note: The sign convention here is different from the
                // 3D case.  Concave-upward curves (smiles) have a positive
                // curvature.  Concave-downward curves (frowns) have a
                // negative curvature.  Be sure to take that into account!
                if (calculate_d2xyz)
                  {
                    const Real numerator   = this->d2xyzdxi2_map[p] * this->normals[p];
                    const Real denominator = this->dxyzdxi_map[p].norm_sq();
                    libmesh_assert_not_equal_to (denominator, 0);
                    curvatures[p] = numerator / denominator;
                  }
```
and in 3D
```c++
                if (calculate_d2xyz)
                  {
                    // Compute curvature using the typical nomenclature
                    // of the first and second fundamental forms.
                    // For reference, see:
                    // 1) http://mathworld.wolfram.com/MeanCurvature.html
                    //    (note -- they are using inward normal)
                    // 2) F.S. Merritt, Mathematics Manual, 1962, McGraw-Hill
                    const Real L  = -this->d2xyzdxi2_map[p]    * this->normals[p];
                    const Real M  = -this->d2xyzdxideta_map[p] * this->normals[p];
                    const Real N  = -this->d2xyzdeta2_map[p]   * this->normals[p];
                    const Real E  =  this->dxyzdxi_map[p].norm_sq();
                    const Real F  =  this->dxyzdxi_map[p]      * this->dxyzdeta_map[p];
                    const Real G  =  this->dxyzdeta_map[p].norm_sq();

                    const Real numerator   = E*N -2.*F*M + G*L;
                    const Real denominator = E*G - F*F;
                    libmesh_assert_not_equal_to (denominator, 0.);
                    curvatures[p] = 0.5*numerator/denominator;
                  }
```
@roystgnr it seems like we should try and make these signs consistent so that code which attempts to be (or just assumes that it is) dimension agnostic isn't silently broken. E.g. I am not currently doing any dimension handling in my surface tension models for laser welding but I am now realizing that I need to be (so I've created https://github.com/idaholab/moose/issues/24461)

**Reply by roystgnr** [reply]
I've got to agree (with a caveat that we should set up a transition path for existing codes; e.g. you have to set the behavior via a global function and we `libmesh_deprecated()` if you never do so) ... but I presume @jwpeterson had some reason for the current behavior.  If it's just "mathematicians are inconsistent", I'd rather be self-consistent than consistent-with-the-literature, but maybe there's something else I haven't thought of.

**Reply by jwpeterson** [reply]
2003 @jwpeterson was not worried about consistent sign conventions!

libMesh/libmesh@ae2271a57fb

The 2D version seems most consistent to me... i.e. if the surface is curved in the direction of the *outward* normal, that should be positive curvature, and vice-versa.

**Reply by lindsayad** [reply]
2003 @jwpeterson didn't have a care in the world

**Reply by lindsayad** [reply]
I've opened https://github.com/libMesh/libmesh/issues/3560

**Reply by alimostafavi24** [reply]
> `curvature as an Auxvariable` you cannot use this (without extra work) for computing the residuals though. Auxvariables do NOT carry derivative information for the Jacobian
> 
> But if you really want to just look at it, in the output, then you ll have to make a new auxkernel, we dont have that one. I can make it too if you have a week or two to wait

I need this to compute the surface tension force.

Do you think the following is the correct approach to calculate the surface tension force?
![Screenshot from 2023-05-24 09-41-44](https://github.com/idaholab/moose/assets/119982603/53e3b7d4-a1cb-4fdf-a46d-fb5971ebcb96)
```
Real
SurfaceTensionLevelset::computeQpResidual()
{

  
 return _sigma * (_second_c[_qp].tr()) * _grad_c[_qp](_component)* _test[_i][_qp];
}

Real
SurfaceTensionLevelset::computeQpJacobian()
{
  return 0.;
}


Real
SurfaceTensionLevelset::computeQpOffDiagJacobian(unsigned jvar)
{
  if (jvar == _grad_c_var)
  {
    return _sigma * (_second_c[_qp].tr()) * _grad_phi[_j][_qp](_component)* _test[_i][_qp];
  }
  //else if (jvar == _second_c_var)
  //{
  //  return _sigma * (_second_phi[_qp].tr()) * _grad_c[_qp](_component)* _test[_i][_qp];
 // }
    return 0.;
}

```

**Reply by alimostafavi24** [reply]
> You may get curvatures from FE computation directly in libMesh: [get_curvatures](https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1FEAbstract.html#a7b50fa327c33945c6fa3bfd2da230cde). As alluded to by @jiangwen84 it does require pre-requesting [get_d2phi](https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1FEGenericBase.html#a2b469a4bf1136ba8fefb1e32c7fae991)

Can you tell me how can I access this from the header or constructor file?

Thanks,

**Reply by alimostafavi24** [reply]
> Take a look at Elin Olsson et.al, JCP 225 (2007) 785–807 paper where he has a formulation for regularized curvature. I suggest introduce a kernel to compute the curvature variable based on that formulation.
> 
> I have implemented it in our internal MOOSE-based codes, and it works well.

I am quite confused here. Is the level set variable (phi in the moose system), the same as signed distance function or signed distance function needs to be computed accordingly?

**Reply by lindsayad** [reply]
> Can you tell me how can I access this from the header or constructor file?

You could do something like this
```c++
SurfaceTensionLevelSet::SurfaceTensionLevelSet(const InputParameters & params),
  IntegratedBC(params),
  _fe_for_curvatures(_assembly.fePhiFace(FEType(_mesh.hasSecondOrderElements() ? SECOND : FIRST,
                                            LAGRANGE)))
{
  _fe_for_curvatures.get_d2phi();
  _fe_for_curvatures.get_curvatures();
}

Real
SurfaceTensionLevelSet::computeQpResidual()
{
  const auto curvature = _fe_for_curvatures.get_curvatures()[_qp];
```
I'm assuming here that you're implementing this with an `IntegratedBC`. If it's something else let me know

**Reply by alimostafavi24** [reply]
My goal is to add the surface tension force as a source term to the Navier Stokes momentum equation and couple it with level set module. (I don't think I need to use IntegratedBC)

I use "c" for level set variable to not confuse it with the shape function. Here is the governing equation for the surface tension force (First I need to calculate the curvature for the level set variable "c" and then find the surface tension force):



<img width="281" alt="Screen Shot 2023-05-25 at 10 48 37 AM" src="https://github.com/idaholab/moose/assets/119982603/84fd7675-ade6-4d11-b346-fae914503ec6">

And here is my header and constructor which gave me wrong results. I just need to get the curvature at the quadrature points and calculates the residual and off-diagonal Jacobian with it. (Since the surface tension force enter the momentum equation the variable is the velocity component and c would be the coupled value) 

```

#pragma once

// MOOSE includes
#include "Kernel.h"

/**
 * Implements the surface tension force by Olsson et. al. (2007).
 */
class SurfaceTensionLevelset : public Kernel
{
public:
  static InputParameters validParams();

  SurfaceTensionLevelset(const InputParameters & parameters);

protected:
  virtual Real computeQpResidual() override;

  virtual Real computeQpJacobian() override;
  virtual Real computeQpOffDiagJacobian(unsigned int jvar);

  //const VariablePhiSecond & _second_phi;
  const VariableValue & _c;
  const VariableGradient & _grad_c;

  const VariableSecond & _second_c;
  const Real _sigma;
    const PostprocessorValue & _epsilon;
  using Kernel::getPostprocessorValue;

  unsigned int _component;
  unsigned int _grad_c_var;
  unsigned int _second_c_var;
};
```

```

// MOOSE includes
#include "SurfaceTensionLevelset.h"

registerMooseObject("LevelSetApp", SurfaceTensionLevelset);

InputParameters
SurfaceTensionLevelset::validParams()
{
  InputParameters params = Kernel::validParams();
  params.addClassDescription("Surface tension force defined by Olsson et. al. (2007).");
  params.addRequiredCoupledVar(
      "c", "The level set variable");
 params.addParam<Real>("sigma", 72.0 , "Coefficient");
 params.addRequiredParam<PostprocessorName>(
      "epsilon", "The epsilon coefficient to be used in the reinitialization calculation.");
  params.addRequiredParam<unsigned int>("component", "number of component (0 = x, 1 = y, 2 = z)");
  return params;
}

SurfaceTensionLevelset::SurfaceTensionLevelset(const InputParameters & parameters) 
  : Kernel(parameters),
    _c(coupledValue("c")),
    _grad_c(coupledGradient("c")),
    _second_c(coupledSecond("c")),
    _sigma(getParam<Real>("sigma")),
    _epsilon(getPostprocessorValue("epsilon")),
    _component(getParam<unsigned int>("component")), 
    _grad_c_var(coupled("c"))
  //\\  _second_c_var(coupled("w"))
    {
}


Real
SurfaceTensionLevelset::computeQpResidual()
{

   Real s = _grad_c[_qp].norm() + std::numeric_limits<Real>::epsilon(); 
 return _sigma * (_second_c[_qp].tr()) * _grad_c[_qp](_component)/s * _test[_i][_qp];
}

Real
SurfaceTensionLevelset::computeQpJacobian()
{
  return 0.;
}


Real
SurfaceTensionLevelset::computeQpOffDiagJacobian(unsigned jvar)
{
     Real s = _grad_c[_qp].norm() + std::numeric_limits<Real>::epsilon(); 

  if (jvar == _grad_c_var)
  {
    return _sigma * (_second_c[_qp].tr()) * _grad_phi[_j][_qp](_component)/s * _test[_i][_qp];
  }
  //else if (jvar == _second_c_var)
  //{
  //  return _sigma * (_second_phi[_qp].tr()) * _grad_c[_qp](_component)* _test[_i][_qp];
 // }
    return 0.;
}


```

**Reply by jwpeterson** [reply]
I think it could work to solve for curvature ("kappa") as a separate coupled variable in the system using the weak form you have shown. But in your code you don't show any kernels used for computing "kappa" in this way, so I'm a bit confused if that is the approach you use? By the way, you are missing a boundary term in the curvature weak form unless the assumption is grad(c).n = 0 on the whole boundary.

**Reply by alimostafavi24** [reply]
I am assuming Neumann BC right now. To me, curvature is a function of c (or phi) only. Why to solve a PDE for it and add a new variable to the system?




![Screenshot from 2023-05-25 15-40-02](https://github.com/idaholab/moose/assets/119982603/78fd0511-a7ab-480c-aa00-5356bec47846)

**Reply by jwpeterson** [reply]
> Why to solve a PDE for it and add a new variable to the system?

Adding a separate variable would be similar to the so-called "split" formulation of the Cahn-Hilliard equation. The advantage of using a split formulation is that you reduce the maximum order of derivatives of c that you need to evaluate. In your case there would be no need to mess with second derivative information at all. To me the nonlinearity (and hence the Jacobian contributions) also seem much simpler to compute if using a separate variable for kappa. The downside is that you have more DOFs of course.

**Reply by alimostafavi24** [reply]
ComputeCurvature2D gets c and computes the curvature.
```

#pragma once

#include "Kernel.h"

class ComputeCurvature2D: public Kernel
{
public:
  ComputeCurvature2D(const InputParameters & parameters);

  static InputParameters validParams();

protected:
  virtual Real computeQpResidual() override;

  virtual Real computeQpJacobian() override;

  const VariableGradient & _grad_c; // c is the order parameter 


  const Real & _epsilon;

};
```


```
#include "ComputeCurvature2D.h"
#include "NS.h"
#include "MooseVariableFE.h"
#include "SystemBase.h"

registerMooseObject("NavierStokesApp", ComputeCurvature2D);

InputParameters
ComputeCurvature2D::validParams()
{
  InputParameters params = Kernel::validParams();
  params.addClassDescription("Adds the surface tension force to the NS equation.");  
 //params.addParam<Real>("sigma", 72.0, "Coefficient");
 // params.addRequiredParam<unsigned int>("component", "number of component (0 = x, 1 = y, 2 = z)");
  params.addRequiredCoupledVar("c", "The order parameter");
 params.addParam<Real>("epsilon", 0.00001, "epsilon");



  return params;
}

ComputeCurvature2D::ComputeCurvature2D(const InputParameters & parameters)
  : Kernel(parameters),

    _grad_c(coupledGradient("c")),
    _epsilon(getParam<Real>("epsilon"))


    {
}

Real
ComputeCurvature2D::computeQpResidual()
{
  Real s = _grad_c[_qp].norm() + std::numeric_limits<Real>::epsilon();
  RealVectorValue n_hat = _grad_c[_qp] / s; 
  return  _test[_i][_qp] * _u[_qp] - n_hat * _grad_test[_i][_qp] ;

}

Real
ComputeCurvature2D::computeQpJacobian()
{
  return _phi[_j][_qp] * _test[_i][_qp];
}
```
This one computes the components of the surface tension force after getting curvature and c.


```
#pragma once

// MOOSE includes
#include "Kernel.h"

/**
 */
class SurfaceTensionLevelsetCurvature : public Kernel
{
public:
  static InputParameters validParams();

  SurfaceTensionLevelsetCurvature(const InputParameters & parameters);

protected:
  virtual Real computeQpResidual() override;

  virtual Real computeQpJacobian() override;
  virtual Real computeQpOffDiagJacobian(unsigned int jvar);

  const VariableValue & _c;
  const VariableGradient & _grad_c;
  const VariableValue & _curvature;

  const Real _sigma;

  unsigned int _component;
  unsigned int _grad_c_var;
  unsigned int _curvature_var;
};
```

```
#include "SurfaceTensionLevelsetCurvature.h"

registerMooseObject("LevelSetApp", SurfaceTensionLevelsetCurvature);

InputParameters
SurfaceTensionLevelsetCurvature::validParams()
{
  InputParameters params = Kernel::validParams();
  params.addClassDescription("computes the surface tension force");
  params.addRequiredCoupledVar(
      "c", "The level set variable to be reinitialized as signed distance function.");
  params.addParam<Real>("sigma", 72.0 , "Coefficient");
  params.addRequiredParam<unsigned int>("component", "number of component (0 = x, 1 = y, 2 = z)");
  return params;
}

SurfaceTensionLevelsetCurvature::SurfaceTensionLevelsetCurvature(const InputParameters & parameters) 
  : Kernel(parameters),
    _c(coupledValue("c")),
    _grad_c(coupledGradient("c")),
    _curvature(coupledValue("c")),
    _sigma(getParam<Real>("sigma")),

    _component(getParam<unsigned int>("component")), 
    _grad_c_var(coupled("c")),
    _curvature_var(coupled("c"))

    {
}


Real
SurfaceTensionLevelsetCurvature::computeQpResidual()
{

 return _sigma * _curvature[_qp] * _grad_c[_qp](_component) * _test[_i][_qp];
}

Real
SurfaceTensionLevelsetCurvature::computeQpJacobian()
{
  return 0.;
}


Real
SurfaceTensionLevelsetCurvature::computeQpOffDiagJacobian(unsigned jvar)
{

  if (jvar == _grad_c_var)
  {
    return _sigma * _curvature[_qp] * _grad_phi[_j][_qp](_component) *_test[_i][_qp];
  }
  else if (jvar == _curvature_var)
  {
    return _sigma * _phi[_j][_qp] * _grad_c[_qp](_component) * _test[_i][_qp];
  }
    return 0.;
}
```
And this is from the input file:

```
  [./curvature]
    type = ComputeCurvature2D

    variable = curvature

    c = c
  [../]
```
```

  [./force_x]
   type = SurfaceTensionLevelsetCurvature
    variable = vel_x
    sigma = ${sigma}
    component = 0    
    c = c
    curvature = curvature
  [../]

  [./force_y]
    type = SurfaceTensionLevelsetCurvature
    variable = vel_y
    sigma = ${sigma}
    component = 1
    c = c
    curvature = curvature
  [../]
```
Something is wrong here and I cannot figure it out.

**Reply by alimostafavi24** [reply]
> ComputeCurvature2D gets c and computes the curvature.
> 
> ```
> 
> #pragma once
> 
> #include "Kernel.h"
> 
> class ComputeCurvature2D: public Kernel
> {
> public:
>   ComputeCurvature2D(const InputParameters & parameters);
> 
>   static InputParameters validParams();
> 
> protected:
>   virtual Real computeQpResidual() override;
> 
>   virtual Real computeQpJacobian() override;
> 
>   const VariableGradient & _grad_c; // c is the order parameter 
> 
> 
>   const Real & _epsilon;
> 
> };
> ```
> 
> ```
> #include "ComputeCurvature2D.h"
> #include "NS.h"
> #include "MooseVariableFE.h"
> #include "SystemBase.h"
> 
> registerMooseObject("NavierStokesApp", ComputeCurvature2D);
> 
> InputParameters
> ComputeCurvature2D::validParams()
> {
>   InputParameters params = Kernel::validParams();
>   params.addClassDescription("Adds the surface tension force to the NS equation.");  
>  //params.addParam<Real>("sigma", 72.0, "Coefficient");
>  // params.addRequiredParam<unsigned int>("component", "number of component (0 = x, 1 = y, 2 = z)");
>   params.addRequiredCoupledVar("c", "The order parameter");
>  params.addParam<Real>("epsilon", 0.00001, "epsilon");
> 
> 
> 
>   return params;
> }
> 
> ComputeCurvature2D::ComputeCurvature2D(const InputParameters & parameters)
>   : Kernel(parameters),
> 
>     _grad_c(coupledGradient("c")),
>     _epsilon(getParam<Real>("epsilon"))
> 
> 
>     {
> }
> 
> Real
> ComputeCurvature2D::computeQpResidual()
> {
>   Real s = _grad_c[_qp].norm() + std::numeric_limits<Real>::epsilon();
>   RealVectorValue n_hat = _grad_c[_qp] / s; 
>   return  _test[_i][_qp] * _u[_qp] - n_hat * _grad_test[_i][_qp] ;
> 
> }
> 
> Real
> ComputeCurvature2D::computeQpJacobian()
> {
>   return _phi[_j][_qp] * _test[_i][_qp];
> }
> ```
> 
> This one computes the components of the surface tension force after getting curvature and c.
> 
> ```
> #pragma once
> 
> // MOOSE includes
> #include "Kernel.h"
> 
> /**
>  */
> class SurfaceTensionLevelsetCurvature : public Kernel
> {
> public:
>   static InputParameters validParams();
> 
>   SurfaceTensionLevelsetCurvature(const InputParameters & parameters);
> 
> protected:
>   virtual Real computeQpResidual() override;
> 
>   virtual Real computeQpJacobian() override;
>   virtual Real computeQpOffDiagJacobian(unsigned int jvar);
> 
>   const VariableValue & _c;
>   const VariableGradient & _grad_c;
>   const VariableValue & _curvature;
> 
>   const Real _sigma;
> 
>   unsigned int _component;
>   unsigned int _grad_c_var;
>   unsigned int _curvature_var;
> };
> ```
> 
> ```
> #include "SurfaceTensionLevelsetCurvature.h"
> 
> registerMooseObject("LevelSetApp", SurfaceTensionLevelsetCurvature);
> 
> InputParameters
> SurfaceTensionLevelsetCurvature::validParams()
> {
>   InputParameters params = Kernel::validParams();
>   params.addClassDescription("computes the surface tension force");
>   params.addRequiredCoupledVar(
>       "c", "The level set variable to be reinitialized as signed distance function.");
>   params.addParam<Real>("sigma", 72.0 , "Coefficient");
>   params.addRequiredParam<unsigned int>("component", "number of component (0 = x, 1 = y, 2 = z)");
>   return params;
> }
> 
> SurfaceTensionLevelsetCurvature::SurfaceTensionLevelsetCurvature(const InputParameters & parameters) 
>   : Kernel(parameters),
>     _c(coupledValue("c")),
>     _grad_c(coupledGradient("c")),
>     _curvature(coupledValue("c")),
>     _sigma(getParam<Real>("sigma")),
> 
>     _component(getParam<unsigned int>("component")), 
>     _grad_c_var(coupled("c")),
>     _curvature_var(coupled("c"))
> 
>     {
> }
> 
> 
> Real
> SurfaceTensionLevelsetCurvature::computeQpResidual()
> {
> 
>  return _sigma * _curvature[_qp] * _grad_c[_qp](_component) * _test[_i][_qp];
> }
> 
> Real
> SurfaceTensionLevelsetCurvature::computeQpJacobian()
> {
>   return 0.;
> }
> 
> 
> Real
> SurfaceTensionLevelsetCurvature::computeQpOffDiagJacobian(unsigned jvar)
> {
> 
>   if (jvar == _grad_c_var)
>   {
>     return _sigma * _curvature[_qp] * _grad_phi[_j][_qp](_component) *_test[_i][_qp];
>   }
>   else if (jvar == _curvature_var)
>   {
>     return _sigma * _phi[_j][_qp] * _grad_c[_qp](_component) * _test[_i][_qp];
>   }
>     return 0.;
> }
> ```
> 
> And this is from the input file:
> 
> ```
>   [./curvature]
>     type = ComputeCurvature2D
> 
>     variable = curvature
> 
>     c = c
>   [../]
> ```
> 
> ```
> 
>   [./force_x]
>    type = SurfaceTensionLevelsetCurvature
>     variable = vel_x
>     sigma = ${sigma}
>     component = 0    
>     c = c
>     curvature = curvature
>   [../]
> 
>   [./force_y]
>     type = SurfaceTensionLevelsetCurvature
>     variable = vel_y
>     sigma = ${sigma}
>     component = 1
>     c = c
>     curvature = curvature
>   [../]
> ```
> 
> Something is wrong here and I cannot figure it out.

The main problem here is that curvature is computed incorrectly. I think there is a mistake in computing the residual for the curvature variable!
